### PR TITLE
[core] Fix bugs in ray syncer about warning and graceful destructing.

### PR DESCRIPTION
**Pull request by fishbone**
Signed-off-by: Yi Cheng <74173148+iycheng@users.noreply.github.com>

## Why are these changes needed?
There are two issues fixed in this PR:

In the current message flow, warning is pushed from raylet to gcs and then to autoscaler. This flow is really bad since it couple a lot of components for a simple feature (raylet/gcs/autoscaler) and ideally it should just be put in autoscaler.

Ray syncer failed in this feature because the state is not carried to autoscaler. To fix this, we'd better fix the flow, but given the complexity, we tried to fix it in syncer.

The new state is carried and pushed to autoscaler as previous.

The other one is about graceful exits. When the structure got shutdown, we need to make sure all init requests exits graceful to avoid crash.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: src/ray/raylet/node_manager.cc [hunk 1]
```diff
@@ -2855,6 +2855,7 @@ std::optional<syncer::RaySyncMessage> NodeManager::CreateSyncMessage(
 
   rpc::ResourcesData resources_data;
   resources_data.set_should_global_gc(true);
+  resources_data.set_cluster_full_of_actors_detected(resource_deadlock_warned_ >= 1);
   syncer::RaySyncMessage msg;
   msg.set_version(absl::GetCurrentTimeNanos());
   msg.set_node_id(self_node_id_.Binary());
```
Comment (jjyao) [inline src/ray/raylet/node_manager.cc:2861]: I feel the code doesn't match the PR description? Which line of code fixes the warning issue?

Comment (iycheng) [reply src/ray/raylet/node_manager.cc:2861]: it matches. The old code works like: https://sourcegraph.com/github.com/ray-project/ray@master/-/blob/src/ray/raylet/node_manager.cc?L1816

btw, it's a bad design. we only want it warn once.

this parameter controls this.

Comment (iycheng) [reply src/ray/raylet/node_manager.cc:2861]: also here fyi, https://sourcegraph.com/github.com/ray-project/ray@master/-/blob/src/ray/raylet/node_manager.cc?L943-945

Comment (jjyao) [reply src/ray/raylet/node_manager.cc:2861]: What about other changes in this PR?

**Review by jjyao — APPROVED** [review]

**Comment by iycheng** [comment]
test failure not related.

**Pull request merged**